src/field/pokemon.ts

```diff
@@ -3061,16 +3094,28 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     const randomMultiplier = simulated ? 1 : this.randSeedIntRange(85, 100) / 100;
 
     const sourceTypes = source.getTypes();
-    const sourceTeraType = source.getTeraType();
+    const sourceTeraType = source.teraType;
     const matchesSourceType = sourceTypes.includes(moveType);
     /** A damage multiplier for when the attack is of the attacker's type and/or Tera type. */
     const stabMultiplier = new NumberHolder(1);
-    if (matchesSourceType) {
+
+    if (matchesSourceType && moveType !== ElementalType.STELLAR) {
       stabMultiplier.value += 0.5;
     }
+
     applyMoveAttrs(CombinedPledgeStabBoostAttr, source, this, move, stabMultiplier);
-    if (sourceTeraType !== ElementalType.UNKNOWN && sourceTeraType === moveType) {
-      stabMultiplier.value += 0.5;
+
+    if (source.isTerastallized) {
+      if (sourceTeraType === moveType && moveType !== ElementalType.STELLAR) {
+        stabMultiplier.value += 0.5;
+      }
+
+      if (
+        source.teraType === ElementalType.STELLAR
+        && (!source.stellarTypesBoosted.includes(moveType) || source.species.speciesId === SpeciesId.TERAPAGOS)
+      ) {
+        stabMultiplier.value += matchesSourceType ? 0.5 : 0.2;
+      }
     }
 
     applyAbFunc<StabBoostAbAttr>(AbAttrFlag.STAB_BOOST, source, simulated, stabMultiplier);
```

Comment:
NIT: Might wanna move the whole `stab` logic into an enclosed helper method (out of scope possibly)